src/runtime/rs.cpp

```diff
@@ -0,0 +1,1213 @@
+        }
+
+        if (loadSymbols(handle, *dispatch) == false) {
+            ALOGV("%s init failed!", filename);
+            return false;
+        }
+        ALOGV("Successfully loaded %s", filename);
+        return true;
+    }
+
+    // Constructor sets 'error' if any occurs.
+    Context(void *user_context)
+        : user_context(user_context), mDev(NULL), mContext(NULL),
+          error(RS_ERROR_RUNTIME_ERROR) {
+#ifdef DEBUG_RUNTIME
+        halide_start_clock(user_context);
+#endif
+        error = halide_rs_acquire_context(user_context, &mDev, &mContext);
+        halide_assert(user_context, mDev != NULL);
+        halide_assert(user_context, mContext != NULL);
+    }
+
+    ~Context() { halide_rs_release_context(user_context); }
+};
+
+WEAK dispatchTable *Context::dispatch;
+
+// Structure to hold the state of a module attached to the context.
+// Also used as a linked-list to keep track of all the different
+// modules that are attached to a context in order to release them all
+// when then context is released.
+struct module_state {
+    void *module;
+    module_state *next;
+};
+WEAK module_state *state_list = NULL;
+
+WEAK int create_rs_context(void *user_context, RsDevice *dev, RsContext *ctx) {
+    uint32_t flags = 0;
+
+    int targetApi = RS_VERSION;
+    Context::dispatch = new dispatchTable;
+
+    bool usingNative = false;
+
+    // attempt to load libRS, load libRSSupport on failure
+    // if property is set, proceed directly to libRSSupport
+    if (getProp("debug.rs.forcecompat") == 0) {
+        usingNative = Context::loadSO("libRS.so");
+    }
+    if (usingNative == false) {
+        if (Context::loadSO("libRSSupport.so") == false) {
+            ALOGV("Failed to load libRS.so and libRSSupport.so");
+            return -1;
+        }
+    }
+
+    *dev = Context::dispatch->DeviceCreate();
+    if (*dev == 0) {
+        ALOGV("Device creation failed");
+        return -1;
+    }
+
+    ALOGV("Created device %p", *dev);
+
+    if (flags &
+        ~(RS_CONTEXT_SYNCHRONOUS | RS_CONTEXT_LOW_LATENCY |
+          RS_CONTEXT_LOW_POWER)) {
+        ALOGE("Invalid flags passed");
+        return -1;
+    }
+
+    *ctx = Context::dispatch->ContextCreate(*dev, 0, targetApi,
+                                            RS_CONTEXT_TYPE_NORMAL, flags);
+    if (*ctx == 0) {
+        ALOGE("Context creation failed");
+        return -1;
+    }
+
+    ALOGV("Created context %p", *ctx);
+    return RS_SUCCESS;
+}
+}
+}
+}
+}  // namespace Halide::Runtime::Internal:RS:
+
+bool loadSymbols(void *handle, dispatchTable &dispatchTab, int device_api) {
+// fucntion to set the native lib path for 64bit compat lib.
+#ifdef __LP64__
+    dispatchTab.SetNativeLibDir =
+        (SetNativeLibDirFnPtr)dlsym(handle, "rsaContextSetNativeLibDir");
+    if (dispatchTab.SetNativeLibDir == NULL) {
+        LOG_API("Couldn't initialize dispatchTab.SetNativeLibDir");
+        return false;
+    }
+#endif
+    dispatchTab.AllocationGetType =
```

Comment:
This chunk of setup code actually also is coming from Android source code base(https://android.googlesource.com/platform/frameworks/rs/+/master/cpp/rsDispatch.cpp), so it might make sense to preserve it as is.